### DISCUSSION TITLE
How to configure port forwarding?

### DISCUSSION THREAD

**Discussion by apbarratt** [nextcloud/all-in-one#674] [Questions] [answered]
I feel like I'm going round in circles here.  No matter what I do, I end up with the same problem telling me that "The server is not reachable on Port 443."

I started with a cloudflare tunnel, gave up on the idea. Was setting up a reverse proxy but figured I didn't actually need it.  So I've gone pure and simple.  I've launched the docker using the command in the readme and I've gone into my router and set up port forwarding, but absolutely nothing works.

The readme would suggest I only need to forward 443 for TCP and 3478 for TCP and UDP.  So that's precisely what I've done.

I've tried pointing WAN 443 to host:443, didn't work.  To host:8443, nope.  To host:8080 (this worked on the tunnel before to at least show the landing page) nope, not this time.
I've tried forwarding 80, 443, 8080, 8443 to their said same ports on the host.
I recall on the tunnel version I had to tell it not to verify TLS (presumably some clash over certificates seeing as cloudflare provides its own) so I do wonder if there's an issue here.

Come on guys, what am I missing?  Surely I'm not the only one trying to set this stuff up on a plain ubuntu box using Cloudflare DNS? :(

**Comment by szaimen** [comment, marked as answer]
Hello, it might be that you are behind CGNAT and thus not able to open port 443. in this case would cloudflare be a possible solution.

Why did you not continue in https://github.com/nextcloud/all-in-one/discussions/655? This looked promising to me...

**Reply by apbarratt** [reply]
Eugh.  Yes, it seems CGNAT is a problem for me.  I didn't know it blocked 443.
I think I'm at the giving up point then.  No solution works if the AIO landing page blocks me from getting any closer.  If it's going to just keep telling me that 443 isn't accessible, then I'm locked out aren't I?

**Reply by szaimen** [reply]
So continuing with https://github.com/nextcloud/all-in-one/discussions/655 did not work out for you?

**Reply by apbarratt** [reply]
My last point on the tunnel solution was successfully making the result of http://localhost:11000 (a random key string) show up at my tunnel end at https://cloud.mydomain.tld
If I then went to https://localhost:8080 and entered https://cloud.mydomain.tld, I got the same error I'd been getting all along: "Domain does not point to this server or reverse proxy not configured correctly."

**Reply by szaimen** [reply]
> If I then went to https://localhost:8080 and entered https://cloud.mydomain.tld, I got the same error I'd been getting all along: "Domain does not point to this server or reverse proxy not configured correctly."

all right. I suppose cloudflare blocked the attempt to connect to the endpoint then. I am thinking about adding an option for these cases where a connection to the endpoint is not possible due to cloudflare blocking or not able to open port 443.

**Reply by apbarratt** [reply]
It certainly would be helpful.  Especially given that the desired security is presumably being provided by cloudflare's systems and their own certificates.

**Reply by apbarratt** [reply]
I'm seeing one further point about using IPV6 to avoid CGNAT issues.  Unfortunately it appears my ISP does not provide me with an IPv6 address.  I think my last remaining option to be able to self hosting options is to find a VPN provider to give me an IP address.  I'd just really hoped to avoid any more third parties as much as possible :(

**Reply by szaimen** [reply]
allright. I've created https://github.com/nextcloud/all-in-one/pull/680 to make debuggin a bit easier. If we come to the conclusion that we need to add a way to work around the last check, I'll add this in the release after the next one. The PR will help in for figuring this out, too.

**Reply by apbarratt** [reply]
What is really confusing me is that I feel like I can't have CGNAT involved here as the IP address reported on my router is the same as the one reported in websites.  I can also connect directly to web servers using port 80 with port forwarding.  It feels like it really is just AIO that's complaining :(

**Reply by szaimen** [reply]
> I can also connect directly to web servers using port 80 with port forwarding. It feels like it really is just AIO that's complaining :(

it needs to connect to port 443 but anyway, we'll improve the verbosity of the domaincheck in the next release with the mentioned PR which hopefully allows to debug the problem more easily.

**Reply by apbarratt** [reply]
But that is precisely the problem.  I *can* reach port 80 and 443.  They're not in the least bit blocked.  I have another server running on the same network with both ports 80 and 443 running servers.  If I direct the address in question to this server, I can see the the web pages running on it.  If I point them at the box running AIO, it sees nothing.
Bypassing cloudflare all together, if I type in my own IP address, I can see port 80 and 443 hosting pages if I point them at my existing server. If I forward them however to the box running AIO, I see nothing.

Throughout all of this, I can't help but feel this makes sense because at no point have I done anything at all to expose port 443 on the box in question.  After all, the original docker command reads:
```
# For x64 CPUs:
sudo docker run -it \
--name nextcloud-aio-mastercontainer \
--restart always \
-p 80:80 \
-p 8080:8080 \
-p 8443:8443 \
--volume nextcloud_aio_mastercontainer:/mnt/docker-aio-config \
--volume /var/run/docker.sock:/var/run/docker.sock:ro \
nextcloud/all-in-one:latest
```
which never exposes 443 at all.

Indeed, nmap confirms all of this:
```
apbarratt@lenovo:~$ sudo nmap localhost
Starting Nmap 7.80 ( https://nmap.org ) at 2022-05-20 12:58 UTC
Nmap scan report for localhost (127.0.0.1)
Host is up (0.000031s latency).
Not shown: 995 closed ports
PORT     STATE SERVICE
22/tcp   open  ssh
80/tcp   open  http
3306/tcp open  mysql
8080/tcp open  http-proxy
8443/tcp open  https-alt

Nmap done: 1 IP address (1 host up) scanned in 0.25 seconds
```
whilst also making me a little nervous wonderig why I've got mysql exposed (must look into that!)

So I suppose my question is, why does AIO's landing page want to check if anything is coming in over port 443 when we never actually exposed anything to port 443?

**Reply by szaimen** [reply]
it checks port 443 because the mastercontainer should start a second container automatically that should use port 443. is the domaincheck container running?
(nextcloud-aio-domaincheck)
The container only gets started as soon as you log in to the aio interface

**Reply by apbarratt** [reply]
I'll poke at docker, see if I can find a list of stuff running :)
Is there a discord or similar available?  I feel like we're getting into support chat that is a bit niche for a github discussion :)

**Reply by szaimen** [reply]
Nope, there is no discord or similar.

**Reply by apbarratt** [reply]
```
apbarratt@lenovo:~$ sudo docker ps
[sudo] password for apbarratt: 
CONTAINER ID   IMAGE                              COMMAND                  CREATED          STATUS          PORTS                                                                                                                     NAMES
218b18d658d3   nextcloud/aio-domaincheck:latest   "/start.sh"              25 minutes ago   Up 25 minutes   0.0.0.0:11000->11000/tcp, :::11000->11000/tcp                                                                             nextcloud-aio-domaincheck
c6738d9dc97d   nextcloud/all-in-one:latest        "start.sh /usr/bin/s…"   18 hours ago     Up 17 hours     0.0.0.0:80->80/tcp, :::80->80/tcp, 0.0.0.0:8080->8080/tcp, :::8080->8080/tcp, 0.0.0.0:8443->8443/tcp, :::8443->8443/tcp   nextcloud-aio-mastercontainer
```

**Reply by apbarratt** [reply]
Curiously, I can see the stuff on port 11000 when I go to http://lenovo.local:11000 but it's not showing up in that nmap list :/

**Reply by szaimen** [reply]
alright, looks good. it seems like the domaincheck container is still listening on port 11000. That is probably the reason why it cannot connect to it (it should be listening on port 443 if you do not want to run it behind a reverse proxy).

**Reply by szaimen** [reply]
> Curiously, I can see the stuff on port 11000 when I go to http://lenovo.local:11000 but it's not showing up in that nmap list :/

I don't know enough about nmap to know why this is the case.

**Reply by apbarratt** [reply]
Hmmm, how can I kill everything and start from scratch?  I've tried it before but I must have left something behind because the original AIO password still applied :/

**Reply by szaimen** [reply]
I am currently working on the documentation for this. See https://github.com/nextcloud/all-in-one/tree/enh/665/reset-instance#how-to-properly-reset-the-instance

**Reply by szaimen** [reply]
okay, this does not work curreently... let me fix that

**Reply by apbarratt** [reply]
I always find this the most frustrating part of docker, so much can be done with a single command (I especially love a good `docker compose up` so I don't have to get headaches figuring it all out each time) but cleaning up to clean slate status is always such a pain!

**Reply by szaimen** [reply]
Now it should work, just fixed the documentation

**Reply by apbarratt** [reply]
Ah, that prune command was nifty.  I thought it was enough to `kill` containers and `rmi -f` images.  I'd forgotten volumes but I was surprised to see it pruning stuff when I thought I'd already killed everything off :)
Right, let's start properly from scratch.  Man I'll be upset if it turns out the solution was just "turn it off and on again" and I just hadn't been doing it right 😅

**Reply by szaimen** [reply]
> Right, let's start properly from scratch. Man I'll be upset if it turns out the solution was just "turn it off and on again" and I just hadn't been doing it right 😅

I really hope that this is the solution :)

**Reply by apbarratt** [reply]
So before I start it up again, what port and protocol should my tunnel point at?
https://localhost:443 or https://localhost:8080 or http://localhost:8080 or https://localhost:11000 or http://localhost:11000
Or if port forwarding, which port to what port?
443 to 443 or to 8443 or to 8080 or 11000...

**Reply by apbarratt** [reply]
My preferred option would be the tunnel, as I want to know that should anything ever go wrong, I can grab the box, run to a friend's house and plug it into their router's ethernet port 😅

**Reply by apbarratt** [reply]
However, I'm also aware tunnels don't allow UDP, so that'll be a challenge to worry about too as I understand the talk thingies need it.

**Reply by szaimen** [reply]
> So before I start it up again, what port and protocol should my tunnel point at?
> https://localhost:443 or https://localhost:8080 or http://localhost:8080 or https://localhost:11000 or http://localhost:11000
> Or if port forwarding, which port to what port?
> 443 to 443 or to 8443 or to 8080 or 11000...

with cloudflare you need to set apache_port to e.g. 11000 and point it at http://localhost:11000

Without you do not set the apache_port and open port 443 in your router which should port forward to port 443 of the server that is running the AIO instance

**Reply by apbarratt** [reply]
Okay.  So tunnel method leads me to the same old issue.  The tunnel points to http://localhost:11000 which only shows a key string.  This can be seen in the browser but the landing page says it's unreachable over port 443.

Resetting again, I went down the route of 443 through port forwarding instead.  This time I can now see the key string when I navigate to http://lenovo.local:443 (note that I'm using http but not https).  There is nothing accessible on the https protocol, despite 443 usually belonging to this protocol.
This itself causes a new problem, because it seems Cloudflare (who I use for DNS also) capture any request to 443 over plain text http protocol, giving a 400 error from cloudflare stating that only https can be used over port 443.

Is the check itself indeed trying to access port 443 over plain text http?

**Reply by szaimen** [reply]
> Okay. So tunnel method leads me to the same old issue. The tunnel points to http://localhost:11000 which only shows a key string. This can be seen in the browser but the landing page says it's unreachable over port 443.

As I said, this is expected. You need to leave it like this and open the AIO interface locally via port 8080.

**Reply by szaimen** [reply]
> Is the check itself indeed trying to access port 443 over plain text http?

yes, it does.

**Reply by apbarratt** [reply]
Then that indeed is where all our problems lie.  As cloudflare it appears will not allow http over 443 through any address they manage :(

**Reply by szaimen** [reply]
It seems like, yes. However in reverse proxy mode it should do the test and the request over https...

**Reply by apbarratt** [reply]
And reverse proxy mode should simply involve me turning on that port 11000 thing on apache?  Because that continues to not work :(
Admitedly the error is different, mentioning a reverse proxy.

**Reply by apbarratt** [reply]
With so many people using cloudflare, surely I'm not the first person to be encountering these issues?  There must be something else that I'm doing wrong if this is the first time this has come up :/

**Reply by szaimen** [reply]
> And reverse proxy mode should simply involve me turning on that port 11000 thing on apache? Because that continues to not work :(

the apache container will use the defined apache_port which should be set to somethine different than 443.
Then you can use it behind a reverse proxy (like cloudflare) which should do the SSL offload.

**Reply by szaimen** [reply]
> With so many people using cloudflare, surely I'm not the first person to be encountering these issues? There must be something else that I'm doing wrong if this is the first time this has come up :/

I think you are probalby the first that tries to use AIO behind cloudflare...

**Reply by apbarratt** [reply]
Okay, so as neither option works right now I'm concentrating on the tunnel solution for now.
To make sure I don't have to keep hunting for the correct command, I've stuck it in a docker-compose.yml file that looks like this:
```
version: "3"
services:
  nextcloud:
    container_name: nextcloud-aio-mastercontainer
    image: nextcloud/all-in-one:latest
    restart: always
    ports:
      - 8080:8080
    environment:
      - APACHE_PORT=11000
    volumes:
      - nextcloud_aio_mastercontainer:/mnt/docker-aio-config
      - /var/run/docker.sock:/var/run/docker.sock:ro
volumes:
  nextcloud_aio_mastercontainer:
```
Which appears to be working.

I've ensured my tunnel is pointing correctly to http://localhost:11000 and figured I might as well give you the address to confirm (I assume if it's a security concern later I can wipe and restart once we know our solution).  So it can be found at https://cloud.barratt.scot

I am then connecting to my box using https://lenovo.local:8080 (I find it most peculiar using ssl over port 8080 but hey ho!) where I can happily login with my password if prompted but again, am faced with the landing page asking for the domain name above but I continue to be told "Domain does not point to this server or reverse proxy not configured correctly."

I can only assume it is still trying to test over http rather than https, despite having set the environment variable to use 11000.

**Reply by szaimen** [reply]
I am slowly starting to think that we indeed go in circles. Lets make a break and continue when https://github.com/nextcloud/all-in-one/pull/680 is merged and at least released to the beta channel. Hopefully debugging will become a bit easier then.

**Reply by apbarratt** [reply]
Sounds good, I go and stick a follow on the PR.

**Reply by apbarratt** [reply]
Hi @szaimen,
I went ahead and copied your code into my container so we could get the logging we were hoping for:
`wget https://raw.githubusercontent.com/nextcloud/all-in-one/3bf0a769c7583159eeb697985174e36684be5ec6/php/src/Data/ConfigurationManager.php`
`sudo docker cp ./ConfigurationManager.php nextcloud-aio-mastercontainer:/var/www/docker-aio/php/src/Data/ConfigurationManager.php`
And now the log we are receiving from line 237 of your commit reads like so:
`[Mon May 23 11:01:40.429170 2022] [php:notice] [pid 154] [client 127.0.0.1:54482] The response of the connection attempt to "https://cloud.barratt.scot:443" was: , referer: https://tiddles.local:8080/containers`
(full disclosure, I got board of the host name "lenovo" and so put it up to a poll within the family, the name "tiddles" was voted in 3 votes against 2)

**Reply by szaimen** [reply]
All right! Thanks for trying!

> [Mon May 23 11:01:40.429170 2022] [php:notice] [pid 154] [client 127.0.0.1:54482] The response of the connection attempt to "https://cloud.barratt.scot:443" was: , referer: https://tiddles.local:8080/containers

Problem is it seems like there was no output logged at all (should have been in between here: `was: ,`)... I hoped that it would log e.g. connection blocked or so but that does not seem to be the case, unfortunately. So that does not really help at debugging the issue...

**Reply by apbarratt** [reply]
A few notes on my network changes.
I have abandoned the cloudflare tunnel and bought myself a VPN subscription with a static IP and open ports on 80, 443, and a phew extras I figured were necessary.
This IP is assigned directly to tiddles rather than messing around with router port forwarding.
I haven't actually turned on any firewall on tiddles, with the assumption that my selective port opening on the VPN side will be enough, but I will probably enable ufw a bit later as it is still of course fully open on my LAN.
The domain is still routed through cloudflare DNS however the proxy side of that is turned off as darling daughter wanted a minecraft server, so tiddles is now running one of those on another port in another docker container.  (I'm getting rather of fond of this docker stuff now!)
I have set up a basic reverse proxy using nginx in another docker container to pass 443 to the correct place (this comment is a bit of a ramble so i'll stick the config for nginx in the next one)
I note that the address in that log is using https, and of course there is nothing being broadcast on that protocol right now, so not sure how we might like to handle that :/

**Reply by apbarratt** [reply]
nginx config:
```
server {
    listen       80;
    server_name  localhost;
    location / {
        root   /usr/share/nginx/html;
        index  index.html index.htm; #To do, stick a pretty landing page or something here or just redirect folk to barratt.scot
    }
}

server {
    listen      443;
    server_name cloud.barratt.scot;
    location / {
        proxy_pass http://192.168.8.3:11000; #the host ip address, localhost I assume would just come back to this docker container
        proxy_set_header X-Real-IP $remote_addr;
        proxy_set_header Host $host;
        proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
        client_max_body_size 0;

        # Websocket
        # proxy_http_version 1.1;
        # proxy_set_header Upgrade $http_upgrade;
        # proxy_set_header Connection $connection_upgrade;
        # I've had to comment this stuff out as I get errors about unknown variable http_upgrade 🤨
    }
}
```

**Reply by apbarratt** [reply]
I need help with the websocket section but I suspect that's a whole other thing so shan't worry on it too much right now.

**Reply by apbarratt** [reply]
I also wondered what would happen if I ran the curl command manually from inside the docker container.  Here are the results for both http and https.
```
apbarratt@tiddles:~$ sudo docker exec -it nextcloud-aio-mastercontainer sh
[sudo] password for apbarratt: 
# curl http://cloud.barratt.scot:443
_big long ID string here as expected_
# curl https://cloud.barratt.scot:443
curl: (35) error:1408F10B:SSL routines:ssl3_get_record:wrong version number
# 
```

**Reply by apbarratt** [reply]
We're definitely out of my knowledge area now 😅
Here is the verbose output of the same command, hopefully it's more informative:
```
# curl https://cloud.barratt.scot:443 -v
*   Trying 109.123.90.42:443...
* Connected to cloud.barratt.scot (109.123.90.42) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* TLSv1.3 (OUT), TLS handshake, Client hello (1):
* error:1408F10B:SSL routines:ssl3_get_record:wrong version number
* Closing connection 0
curl: (35) error:1408F10B:SSL routines:ssl3_get_record:wrong version number
# 
```

**Reply by szaimen** [reply]
Looks like this might help? https://stackoverflow.com/a/62272517

**Reply by apbarratt** [reply]
Sorry! I was summoned to go dig a bunch of holes in the garden!  Trying now...

Right, that upsets things as of course i don't actually have a certificate set up on my server and now it's trying to find one, having relied on the one from cloudflare up until now.  I'll see what the nextcloud proxy docs say about certificates etc, probably just sort out a lets encrypt thingy.

**Reply by apbarratt** [reply]
Took me a wee while to get that one figured out but seems to now all be working :) I just had to be sure I had a real SSL running on my machine for the proxy mode to work correctly.
I'll be sure to make a cleaned up docker compose file to share somewhere for anyone else who might fancy something with nginx and let's encrypt all bundled together :)

**Reply by apbarratt** [reply]
I don't really know why though, but it's going extremely slowly.  This was before it started working.  Like, every time I spin it up from scratch, the AIO landing page would take a good minute or three to actually load up when it wasn't a problem before.  Even when I don't have the other stuff running :/

**Reply by szaimen** [reply]
> Took me a wee while to get that one figured out but seems to now all be working :) I just had to be sure I had a real SSL running on my machine for the proxy mode to work correctly.
> I'll be sure to make a cleaned up docker compose file to share somewhere for anyone else who might fancy something with nginx and let's encrypt all bundled together :)

Congratulations, finally! 🎉

Feel free to add your solution to this thread: https://github.com/nextcloud/all-in-one/discussions/588

**Reply by szaimen** [reply]
> I don't really know why though, but it's going extremely slowly. This was before it started working. Like, every time I spin it up from scratch, the AIO landing page would take a good minute or three to actually load up when it wasn't a problem before. Even when I don't have the other stuff running :/

there were some reports that this might be due to some firewall problems. see E.g. https://github.com/nextcloud/all-in-one/discussions/580 and https://github.com/nextcloud/all-in-one/discussions/538

**Reply by apbarratt** [reply]
Hmmm, yes, I shall definetely see if nginx or similar may be the culprit here but given that even connecting directly to the port 8080 admin console for AIO can be incredibly slow I suspect it's not that.  Possibly I just need to reboot the host machine after days of brutal prodding!
One thing I've found odd though now is that I've got no office stuff to be seen in the nextcloud interface.  I was sure to install the collabora stuff when prompted during AIO setup and had been looking forward to being able to ditch google docs but there's just no sign of it in the files page or anywhere else.  I tried uploading a word document but clicking it only triggers a download.  Most odd.

**Reply by szaimen** [reply]
Hm... collabora  might not work if websockets do not work...

Also what are the specs of the device where you are running it on?

**Reply by apbarratt** [reply]
Ah that would make sense, I'm still to get the websockets side sorted out on nginx.  My guess is that's why the weather stuff isn't loading either.
It's an older Lenovo Thinkstation (but not horrifically so) that I acquired for free.  I forget the precise specs but it's a quad core Pentium (I've no idea what pentium even means these days!) with I think 4GB of ram (must upgrade that!) and the primary disk is a 250GB SSD that I've stuck in there.  First thing I did was give it some disks and stuck Ubuntu 20.04 on it.  I've also got a 1TB hard disk in there that I was hoping to set up as the main location for the files on the cloud but I'm not sure how wise this would be in terms of speed.  Though I suppose given it'll always be limited to network speeds a hard drive would probably keep up.

**Reply by szaimen** [reply]
Yes, websocket support is pretty important...

The specs sound fine!

**Reply by apbarratt** [reply]
Nextcloud reports it as such:
```
Operating System: | Linux 5.4.0-110-generic x86_64
CPU: | Intel(R) Pentium(R) CPU J2900 @ 2.41GHz (4 cores)
Memory: | 3.73 GB
```
Curious that it's not reporting a full 4GB of ram there, I'd expect that from a 32bit system but not 64 :/
Interesting seeing that it's able to get that information even from within the docker container too.

**Reply by apbarratt** [reply]
Whelp, it's 22:38 here, so I'm off to bed.
Thank you for all your help these last couple of days.  Hopefully I'll figure out the websockets proxying tomorrow and let you go be someone else's saviour 😄

**Reply by szaimen** [reply]
> Curious that it's not reporting a full 4GB of ram there, 

That is due to the different counting of 1000 vs 1024...

**Reply by szaimen** [reply]
> Hopefully I'll figure out the websockets proxying tomorrow

All right! Good luck!

**Reply by apbarratt** [reply]
Righto, got the websockets sorted on nginx and after a full reboot on tiddles, I'm all in order.  I'm now very happy with my set up :)
Thank you again, I'll be sure to write up some notes and see if there's anywhere in the docs I could supplement based on them for outsiders like me to get going quicker :)

**Comment by apbarratt** [comment]
I'm at an utter loss for what's going on.  It was working perfectly for a short while there and then, it just stopped.  I've tried killing everything, restarting the box, whatever I can think of.  But I'm faced simply with the fact that now the only things running are my nginx and nextcloud-aio-mastercontainer containers but absolutely nothing is loading on https://tiddles.local:8080
It simply starts the request and then just sort of hangs there, trying to load.  Sometimes I might see the containers page load up after a great many minutes of waiting but just as unresponsive, most times just nothing happens.

I've tried killing all the containers, pruning them all, etc, it's just dead and I cannot for the life of me think what it is I've done to cause this :/

**Reply by szaimen** [reply]
lets continue in https://github.com/nextcloud/all-in-one/discussions/730
